File: backend/domain/src/main/java/com/swiftly/domain/Vehicle.java
package com.swiftly.domain;

import com.swiftly.domain.enums.vehicle.Feature;
import com.swiftly.domain.enums.vehicle.FuelType;
import com.swiftly.domain.enums.vehicle.VehicleType;
import jakarta.persistence.*;
import lombok.*;

import java.time.Instant;
import java.util.List;


@NoArgsConstructor
@Getter
@Setter
@AllArgsConstructor

public class Vehicle {

    private Integer id;

    private User owner;

    private String vin;

    private String make;

    private String model;

    private String color;

    private Integer year;

    private VehicleType type;

    private FuelType fuelType;

    private Double fuelConsumption;

    private List<Feature> features;

    private String country;

    private String city;

    private Double latitude;

    private Double longitude;

    private Instant locationTimeStamp;

    private List<VehicleImage> images;



    public Vehicle (String vin, String make, String model, String color, Integer year, VehicleType type, FuelType fuelType, Double fuelConsumption, List<Feature> features, String country, String city)
    {
        this.vin = vin;
        this.make = make;
        this.model = model;
        this.color = color;
        this.year = year;
        this.type = type;
        this.fuelType = fuelType;
        this.fuelConsumption = fuelConsumption;
        this.features = features;
        this.country = country;
        this.city = city;
    }

    public Vehicle (Integer id,User owner, String vin, String make, String model, String color, Integer year, VehicleType type, FuelType fuelType, Double fuelConsumption, List<Feature> features, String country, String city)
    {
        this.id = id;
        this.owner = owner;
        this.vin = vin;
        this.make = make;
        this.model = model;
        this.color = color;
        this.year = year;
        this.type = type;
        this.fuelType = fuelType;
        this.fuelConsumption = fuelConsumption;
        this.features = features;
        this.country = country;
        this.city = city;
    }

    public Vehicle (User owner, String vin, String make, String model, String color, Integer year, VehicleType type, FuelType fuelType, Double fuelConsumption, List<Feature> features, String country, String city)
    {
        this.owner = owner;
        this.vin = vin;
        this.make = make;
        this.model = model;
        this.color = color;
        this.year = year;
        this.type = type;
        this.fuelType = fuelType;
        this.fuelConsumption = fuelConsumption;
        this.features = features;
        this.country = country;
        this.city = city;
    }

    public Vehicle (User owner, String vin, String make, String model, String color, Integer year, VehicleType type, FuelType fuelType, Double fuelConsumption, List<Feature> features, String country, String city, List<VehicleImage> images)
    {
        this.owner = owner;
        this.vin = vin;
        this.make = make;
        this.model = model;
        this.color = color;
        this.year = year;
        this.type = type;
        this.fuelType = fuelType;
        this.fuelConsumption = fuelConsumption;
        this.features = features;
        this.country = country;
        this.city = city;
        this.images = images;
    }

    public Vehicle (Integer id,User owner, String vin, String make, String model, String color, Integer year, VehicleType type, FuelType fuelType, Double fuelConsumption, List<Feature> features, String country, String city, List<VehicleImage> images)
    {
        this.id = id;
        this.owner = owner;
        this.vin = vin;
        this.make = make;
        this.model = model;
        this.color = color;
        this.year = year;
        this.type = type;
        this.fuelType = fuelType;
        this.fuelConsumption = fuelConsumption;
        this.features = features;
        this.country = country;
        this.city = city;
        this.images = images;
    }
}

File: frontend/src/types/vehicle.ts
import type { SelectMenuOption } from "@/lib/types";
import type {
  VehicleSchemaType,
  VehicleUpdateSchemaType,
} from "@/schemas/vehicle/vehicle.schema";
import type { UseFormReturn } from "react-hook-form/dist/types/form";
import type { Profile } from "./profile";

export interface AddVehicleFormProps {
  addVehicleForm: UseFormReturn<VehicleSchemaType | VehicleUpdateSchemaType>;
  handleSubmit: (data: VehicleSchemaType | VehicleUpdateSchemaType) => void;
  isPending: boolean;
  vehicleTypesOptions: { value: string; label: string }[];
  fuelTypesOptions: { value: string; label: string }[];
  featuresOptions: { value: string; label: string }[];
  countryOpen: boolean;
  setCountryOpen: (open: boolean) => void;
  selectedCountry: SelectMenuOption | null;
  onDelete?: () => void;
  isDeleting?: boolean;
}

export interface VehicleImage {
  id?: number;
  fileName?: string;
  mimeType?: string;
  uploadedAt?: string;
  data?: string;
  [key: string]: unknown;
}

export interface Vehicle {
  id?: number;
  owner?: Profile;
  vin: string;
  make: string;
  model: string;
  color?: string;
  year?: number;
  type?: string;
  fuelType?: string;
  fuelConsumption?: number;
  features?: string[];
  country?: string;
  city?: string;
  listed?: boolean;
  isRemoved?: boolean;
  images?: VehicleImage[];
  [key: string]: unknown;
}

export interface VehicleTableProps {
  vehicles: Vehicle[];
  onRowClick?: (vehicle: Vehicle) => void;
  maxHeight?: string;
  itemsPerPage?: number;
}
